### DISCUSSION TITLE
C extension modules, circular dependencies

### DISCUSSION THREAD

**Discussion by jpfeuffer** [microsoft/pyright#4360] [Q&A] [answered]
Hi, we have a package that consists of Cython bindings to a large C++ lib. The bindings are split into multiple submodules due to its size. The number of submodules depends on how strong the build machine is (parallel compilation). The submodules have dependencies between each other, potentially circular.
For runtime python the following works fine:
- create an _allmodules.py that imports all the symbols from the private submodules (_submodule_1..N.so/pyd)
- in `__init__.py` import everything from _allmodules so the symbols are available as mypackage.symbol

Now I am creating typestubs for the submodules (_submodule_1..N.pyi). I am adding a py.typed as well.
But somehow I cannot get pyright to find symbols of e.g., _submodule_2.pyi in _submodule_1.pyi. Types from other submodules are always Unknown to pyright. I tried many combinations of relative and absolute imports in the submodules stub pyis. I also tried creating the same __init__.pyi, _allmodules.pyi import hierarchy to no avail.

I am using `--verifytypes` on our package. The package name is `pyopenms` if you want to try yourself. You will need a nightly wheel from GitHub.

Questions:
- Can you, from the description, say what would be the theoretically correct way of importing?
- Can I somehow debug **why** a symbol is Unknown or not found? E.g. where did it look for the symbol.

Thanks a lot!

**Comment by erictraut** [comment, marked as answer]
If I understand you correctly, you're creating an `__init__.pyi` stub. Within that stub, you are importing symbols from other stub files such as `_submodule_1.pyi`, and your intent is for these symbols to be re-exported from the top-level module. By default, symbols imported by a stub file are considered private to that stub and not re-exported. To make them visible externally as public (re)exports, you need to either include them in an `__all__` definition or use redundant forms of `import` such as `from _submodule_1 import Foo as Foo`. For more details, refer to PEP 484 or [this documentation](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md).

**Reply by jpfeuffer** [reply]
Thank you, yes I would need to re-export those symbols. How about star imports? (I could not find a mention in your documentation). It would be very tedious to list all thousands of symbols and keep that list up-to-date with the C++ library.

**Reply by jpfeuffer** [reply]
If it follows [484](https://peps.python.org/pep-0484/#:~:text=However%2C%20as%20an%20exception%20to%20the%20previous%20bullet%2C%20all%20objects%20imported%20into%20a%20stub%20using%20from%20...%20import%20*%20are%20considered%20exported.%20(This%20makes%20it%20easier%20to%20re%2Dexport%20all%20objects%20from%20a%20given%20module%20that%20may%20vary%20by%20Python%20version.)), star imports should export everything without and underscore, correct?

**Reply by erictraut** [reply]
Yes, wildcard (star) imports also work, but I recommend against using them. That will pull in all module-level symbols from the import target including global variables like TypeVars and other symbols imported by your submodule, which you probably don't intend to re-export as part of your library's public interface contract. It's better to be explicit about what symbols you intend your library to export.

**Comment by jpfeuffer** [comment]
By the way, I found some errors in my types (such as listing unicode as a possible type in python3) which might contribute to Unknown types but as mentioned in the beginning, I find it very hard to locate the *source* (aka first occurrence) of the type error since Unknown somehow infects everything only even closely related to a type.
I feel like if a single type in a Union of an argument type of a function in a class is not correctly typed, suddenly the whole class "will not be accepted"/"is unknown" to pyright. Is that intended behaviour? And can I maybe loosen this restriction to debug better?

**Reply by erictraut** [reply]
Yes, this is intended behavior. A single unknown type will "infect" all types that depend on it, making them "partially unknown". That's the nature of type systems, and it's why it's so important for library authors to be thorough when providing type information. On behalf of your library consumers, I want to than you for the work you're doing here!

If the source of the Unknown is coming from within your library, it should be pretty easy to track down. If it's coming from types imported from other libraries, then it becomes more difficult. You can use the `--ignoreexternal` command-line option in conjunction with `--verifytypes` to ignore symbols imported from other libraries when debugging your own library's type information.
